### PR TITLE
Fix multiple temperature scaling in sample_non_deterministic()

### PR DESCRIPTION
**Pull request by dansuissa**
Ensures that temperature scaling is applied only once by moving it inside softmax() instead of modifying logits multiple times. Fixes issue #19.

### CODE REVIEW

**Comment by Adibvafa** [comment]
LGTM
